binder/src/main/java/io/grpc/binder/internal/BinderServer.java

```diff
@@ -0,0 +1,165 @@
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A gRPC InternalServer which accepts connections via a host AndroidService.
+ *
+ * <p>Multiple incoming connections transports may be active at a time.
+ *
+ * <b>IMPORTANT</b>: This implementation must comply with this published wire format.
+ * https://github.com/grpc/proposal/blob/master/L73-java-binderchannel/wireformat.md
+ */
+@ThreadSafe
+final class BinderServer implements InternalServer, LeakSafeOneWayBinder.TransactionHandler {
+
+  private final boolean useSharedTimer;
+  private final ScheduledExecutorService executorService;
+  private final ImmutableList<ServerStreamTracer.Factory> streamTracerFactories;
+  private final AndroidComponentAddress listenAddress;
+  private final LeakSafeOneWayBinder hostServiceBinder;
+  private final ServerSecurityPolicy serverSecurityPolicy;
+  private final InboundParcelablePolicy inboundParcelablePolicy;
+
+  @GuardedBy("this")
+  private ServerListener listener;
+
+  @GuardedBy("this")
+  private boolean shutdown;
+
+  BinderServer(
+      AndroidComponentAddress listenAddress,
+      @Nullable ScheduledExecutorService executorService,
+      List<? extends ServerStreamTracer.Factory> streamTracerFactories,
+      ServerSecurityPolicy serverSecurityPolicy,
+      InboundParcelablePolicy inboundParcelablePolicy) {
+    this.listenAddress = listenAddress;
+    useSharedTimer = executorService == null;
+    this.executorService =
+        useSharedTimer ? SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE) : executorService;
+    this.streamTracerFactories =
+        ImmutableList.copyOf(checkNotNull(streamTracerFactories, "streamTracerFactories"));
+    this.serverSecurityPolicy = checkNotNull(serverSecurityPolicy, "serverSecurityPolicy");
+    this.inboundParcelablePolicy = inboundParcelablePolicy;
+    hostServiceBinder = new LeakSafeOneWayBinder(this);
+  }
+
+  /** Return the binder we're listening on. */
+  public IBinder getHostBinder() {
+    return hostServiceBinder;
+  }
+
+  @Override
+  public synchronized void start(ServerListener serverListener) throws IOException {
+    this.listener = serverListener;
+  }
+
+  @Override
+  public SocketAddress getListenSocketAddress() {
+    return listenAddress;
+  }
+
+  @Override
+  public List<? extends SocketAddress> getListenSocketAddresses() {
+    return ImmutableList.of(listenAddress);
+  }
+
+  @Override
+  public InternalInstrumented<SocketStats> getListenSocketStats() {
+    return null;
+  }
+
+  @Override
+  @Nullable
+  public List<InternalInstrumented<SocketStats>> getListenSocketStatsList() {
+    return null;
+  }
+
+  @Override
+  public synchronized void shutdown() {
+    if (!shutdown) {
+      shutdown = true;
+      if (useSharedTimer) {
+        SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, executorService);
```

Comment:
Oh good point, this was clearly a misunderstanding on my part.

It looks like I handle this _better_, but still not correctly within OnDeviceServer.

I've added a TODO for now, and will follow up with a fix. I need to make a couple of changes to this class in a follow up anyway, to support the way we're planning to do OnDeviceServer in the gRFC. I wanted that to be a separate CL since this is largely a copy of the internal codebase.